File: app/src/main/java/org/stepic/droid/model/Video.java
package org.stepic.droid.model;

import android.os.Parcel;
import android.os.Parcelable;

import org.stepic.droid.base.App;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

public class Video implements Parcelable, Serializable {
    private long id;
    private String thumbnail;

    private List<VideoUrl> urls;

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(this.id);
        dest.writeString(this.thumbnail);
        dest.writeList(this.urls);
    }

    public Video() {
    }

    public void setId(long id) {
        this.id = id;
    }

    public void setThumbnail(String thumbnail) {
        this.thumbnail = thumbnail;
    }

    public void setUrls(List<VideoUrl> urls) {
        this.urls = urls;
    }

    protected Video(Parcel in) {
        this.id = in.readLong();
        this.thumbnail = in.readString();
        this.urls = new ArrayList<>();
        in.readList(this.urls, App.Companion.getAppContext().getClassLoader());
    }

    public static final Creator<Video> CREATOR = new Creator<Video>() {
        public Video createFromParcel(Parcel source) {
            return new Video(source);
        }

        public Video[] newArray(int size) {
            return new Video[size];
        }
    };

    public List<VideoUrl> getUrls() {
        return urls;
    }

    public long getId() {
        return id;
    }

    public String getThumbnail() {
        return thumbnail;
    }
}

File: model/src/main/java/org/stepik/android/model/Video.kt
package org.stepik.android.model

import android.os.Parcel
import android.os.Parcelable
import com.google.gson.annotations.SerializedName

data class Video(
    @SerializedName("id")
    val id: Long = 0,
    @SerializedName("thumbnail")
    val thumbnail: String? = null,
    @SerializedName("urls")
    val urls: List<VideoUrl> = emptyList(),
    @SerializedName("duration")
    val duration: Long = 0
) : Parcelable {
    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeLong(id)
        parcel.writeString(thumbnail)
        parcel.writeTypedList(urls)
        parcel.writeLong(duration)
    }

    override fun describeContents(): Int = 0

    companion object CREATOR : Parcelable.Creator<Video> {
        override fun createFromParcel(parcel: Parcel) =
            Video(
                parcel.readLong(),
                parcel.readString(),
                parcel.createTypedArrayList(VideoUrl),
                parcel.readLong()
            )

        override fun newArray(size: Int): Array<Video?> =
            arrayOfNulls(size)
    }
}


data class VideoUrl(
    @SerializedName("url")
    val url: String?,
    @SerializedName("quality")
    val quality: String?
) : Parcelable, Comparable<VideoUrl> {
    override fun compareTo(other: VideoUrl): Int =
        when {
            quality == null && other.quality == null ->
                0
            quality == null ->
                -1
            other.quality == null ->
                1
            else -> {
                val lengthComparison = quality.length.compareTo(other.quality.length)
                if (lengthComparison == 0) {
                    quality.compareTo(other.quality)
                } else {
                    lengthComparison
                }
            }
        }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(url)
        parcel.writeString(quality)
    }

    override fun describeContents(): Int = 0

    companion object CREATOR : Parcelable.Creator<VideoUrl> {
        override fun createFromParcel(parcel: Parcel) =
            VideoUrl(parcel.readString(), parcel.readString())

        override fun newArray(size: Int): Array<VideoUrl?> =
            arrayOfNulls(size)
    }
}